Rotation from euler

adbugger opened the pull request:
Initializing rotations from euler angles.

adbugger committed:
ENH : Initial implementation of `from_euler`

Issues still need to be worked out in input format validation.
Implementation is still returning nans.

adbugger committed:
ENH: Implementation of `from_euler` function

Fixed errors and made a few decisions. Now, in the case of a single
rotations, the user can specify upto 3 angles and axes. In the case
of multiple rotations, however, numpy.asarray requires that the
dimensions match, so there the user must input a (N, 3) matrix only.

adbugger commented:
In the case of multiple rotations, the user must specify all three angles for every rotation. This is because of `np.asarray`. Another approach I see is to let the angles be specified as a list of lists. What do you guys prefer?

For single rotations, up to 3 angles can be specified.

adbugger commented:
The CircleCI pypy3 build had already failed. Since the job was not re-triggered, that tells us nothing new.

adbugger commented:
So for the list of lists approach, `numpy` is able to make an `numpy.ndarray` out of it, with each element of type `list`. However, we lose `numpy` functions such as `np.deg2rad` for conversion and other things.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -281,3 +281,115 @@ def as_rotvec(self):
+        Parameters
+        ----------
+        seq : str or array of str
+            Each str[i] (or str itself) must be a string of upto 3 characters
```
I vote for `seq` being a string only. Typically, if we want to initialize multiple rotations the sequence will be the same for all of them. This will simplify things greatly.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -281,3 +281,115 @@ def as_rotvec(self):
+            intrinsic rotation. Extrinsic and intrinsic rotations cannot be
+            mixed in one function call. Its length must match that of the
+            `angles` parameter.
+        angle_mat : array_like, shape ([1 or 2 or 3], ) or (N, 3)
```
Name it just `angles`. I suggest that we can use it like `from_euler('x', 1)` or `from_euler('x', [1, 2, 3])` as well.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -281,3 +281,115 @@ def as_rotvec(self):
+            Euler angles specified in radians (`deg` is False) or degrees if
+            parameter `deg` is True. Each `angles[i]` corresponds to an array
+            of angles matching the number of axes specified.
+        deg : boolean
```
Perhaps `degrees`.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -281,3 +281,115 @@ def as_rotvec(self):
+                [0, sinx, cosx]
+                ])
+
+        return dcm.transpose() if intrinsic else dcm
```
For extrinsic/intrinsic rotations we don't need to transpose individual matrices, we only reverse the multiplication order. Think about it.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -281,3 +281,115 @@ def as_rotvec(self):
+        return dcm.transpose() if intrinsic else dcm
+
+    @classmethod
+    def _make_dcm_from_euler(cls, axis_str, angle):
```
I'm not sure if I like the approach you took here. I see the two possibilities:

1. We multiply "elementary" quaternions. In this case you need to implement quaternion multiplication routine. You can keep it as a standalone function.
2. We compute dcms in terms of trigonometric functions directly, as explained in the paper I will send to you and then convert to quaternions. Also note that the paper contains an alternative approach to extract Euler angles for an arbitrary sequence. You should definitely study it as well.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -281,3 +281,115 @@ def as_rotvec(self):
+        extrinsic = recompile("^[xyz]{1,3}$")
+        intrinsic = recompile("^[XYZ]{1,3}$")
+
+        if len(axis_str) != angle.shape[0]:
```
It seems to me that you don't check for invalid sequences. I can't claim that we agreed on that 100%, but it looks logical as we decided to stick to classic Euler angles. In this case we should allow sequence where the second axis is not equal to the first and the third axes.

nmayorov reviewed (commented):
This conversion procedure is much more controversial than the previous ones.

adbugger committed:
MAINT: Updated str-language specification

adbugger committed:
ENH: `from_euler` implementation using elementary quaternions

Introduced standalone functions for quaternion composition.

adbugger committed:
TST: Simple tests for `from_euler`

Initial tests. More to be added.

adbugger committed:
STY: Minor change to test file

adbugger commented:
Implemented `from_euler` with composition of elementary quaternions. Added some simple tests, but tests are failing as they are not equal within tolerance. Will try to fix those next.

So there is a difference of `1e-16` in `test_from_euler_single_rotation`. Why is this test failing when `rtol = 1e-7 > 1e-16`? And how should I go about fixing it?

Setting `atol=1e-15` makes the tests pass, but I'm not sure if that's right way to go about it.

adbugger committed:
TST: Changed test to compare quaternions. All tests pass now.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -2,7 +2,7 @@
 import numpy as np
 import scipy.linalg
-import warnings
+from re import compile as recompile  # compile itself is a builtin function
```
Well, just use `import re`.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -281,3 +282,147 @@ def as_rotvec(self):
         else:
             return rotvec
+
+    @classmethod
```
Do we need it as classmethod? If not, then I suggest to move it out from the class completely.

nmayorov replied:
You will make it a standalone function, right?

adbugger replied:
It's not clear from the limited context here but the function just after `as_rotvec` is now `_elementary_quat_compose`. Given that this function accepts information about `seq`, `angles` and the kind of rotation (`intrinsic` / `extrinsic`), does it make sense to move this logic out of the class? Just trying to understand your thought process here.

nmayorov replied:
You can mark it as `@staticmethod`, right? But I don't see much value in having static methods, especially non-public ones. It comes down to preferences, but I think keeping the class code shorter is a good thing. Additionally, you already move out some "static methods", why make an exception here. 

I don't insist on changing it, you can mark it staticmethod.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -281,3 +282,147 @@ def as_rotvec(self):
             return rotvec
+
+    @classmethod
+    def _make_elementary_quat(cls, char, angles):
```
char -> axis?

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -281,3 +282,147 @@ def as_rotvec(self):
+
+    @classmethod
+    def _make_elementary_quat(cls, char, angles):
+        if len(char) != 1:
```
My suggestion is to remove such checks from utility functions. It means that you should prepare/check everything in user API functions or methods before calling it.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -281,3 +282,147 @@ def as_rotvec(self):
+        return quat
+
+    @classmethod
+    def _compose_quat(cls, p, q):
```
Same notes as for the previous function.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -281,3 +282,147 @@ def as_rotvec(self):
+
+        num_quat = p.shape[0]
+        product = np.empty((num_quat, 4))
+        product[:, 0] = (p[:, 3] * q[:, 0] + p[:, 0] * q[:, 3] +
```
Use `np.cross`.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -281,3 +282,147 @@ def as_rotvec(self):
+    def _elementary_quat_compose(cls, seq, angles, intrinsic=False):
+        seq = seq.lower()
+        num_rotations = angles.shape[0]
+        quat = np.empty((num_rotations, 4, len(seq)))
```
If I'm not mistaken this array is not necessary. You can generate elementary quaternions on the fly.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -281,3 +282,147 @@ def as_rotvec(self):
+            raise ValueError("Expected axis specification to be a non-empty "
+                             "string of upto 3 characters, got {}".format(seq))
+
+        if ((len(seq) == 2 and seq[0] == seq[1]) or
```
How about `if any(seq[i] == seq[i + 1] for i in range(len(seq) - 1))`

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -281,3 +282,147 @@ def as_rotvec(self):
+
+        intrinsic = (recompile('^[XYZ]{1,3}$').match(seq) is not None)
+        extrinsic = (recompile('^[xyz]{1,3}$').match(seq) is not None)
+        if not (intrinsic or extrinsic):
```
For me it makes sense to move this check before the previous one.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -281,3 +282,147 @@ def as_rotvec(self):
+        intrinsic = (recompile('^[XYZ]{1,3}$').match(seq) is not None)
+        extrinsic = (recompile('^[xyz]{1,3}$').match(seq) is not None)
+        if not (intrinsic or extrinsic):
+            raise ValueError("Expected either intrinsic or extrinsic rotations"
```
I suggest to be more specific --- "Expected axes from `seq` to be from ['x', 'y', 'z'] or ['X', 'Y, 'Z'], got ...". Or something like that.

nmayorov replied:
I think after all checks are done and intrinsic flag is set we can do `seq = seq.lower()`.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -281,3 +282,147 @@ def as_rotvec(self):
+            seq = seq.lower()
+            if angles.ndim == 0:
+                # Single value case
+                quat = cls._make_elementary_quat(seq, angles.reshape((1)))
```
Just `angles.reshape(1)`.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -281,3 +282,147 @@ def as_rotvec(self):
+            elif angles.ndim == 1:
+                quat = cls._make_elementary_quat(seq, angles)
+            else:
+                raise ValueError("Expected 1D or 2D `angles` parameter for a"
```
I guess "expected float or 1D array?"

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -281,3 +282,147 @@ def as_rotvec(self):
+            If True, then the given angles are taken to be in degrees
+        """
+
+        if len(seq) < 1 or len(seq) > 3:
```
Would it make sense to introduces `n_axes = len(seq)` or `num_axes = len(seq)`?

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -281,3 +282,147 @@ def as_rotvec(self):
+                raise ValueError("Expected 1D or 2D `angles` parameter for a"
+                                 "single character `seq`, "
+                                 "got shape {}".format(angles.shape))
+        else:  # 2 or 3 axes
```
I think it's better to combine both cases into one. To do that we need to add extra dimensions to `angles` in case of one axis if necessary. I think it will make code easier to understand.

So the algorithms would look like this:

1. Prepare `angles` to have shape `(N, n_axes)`.
2. Call the `compose_elementary_rotations`.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -281,3 +282,147 @@ def as_rotvec(self):
+            quat[:, :, idx] = cls._make_elementary_quat(axis, angles[:, idx])
+
+        # Initialize result to identity quaternions
+        result = np.zeros((num_rotations, 4))
```
Not sure if it is better or worse, but you can initialize the quaternions from the first axis and compute products for remaining axes. This way the case of a single axis will work the same way as it is implemented now.

nmayorov reviewed (commented):
Another round of suggestions.

adbugger committed:
MAINT: Changed implementation of `from_euler`.

Implementation of `from_euler` has been changed according to
suggestions. Standalone functions, where appropriate, have been moved
out of the class.

adbugger committed:
TST: Add test case to check order of rotations

Intrinsic rotations should be same as extrinsic with the order of axes
reversed

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -2,7 +2,32 @@
+    # Scalar part of result
+    # Einsum calculates row wise dot product
+    product[:, 3] = (p[:, 3] * q[:, 3] -
+                     np.einsum('ij,ij->i', p[:, :3], q[:, :3]))
```
Not really sure about that, but `np.sum(p[:, :3] * q[:, :3], axis=1)` might worth considering as a simpler approach, might be faster. If you want, you can make a benchmark to find out.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -281,3 +307,113 @@ def as_rotvec(self):
+    def _elementary_quat_compose(cls, seq, angles, intrinsic=False):
+        # Initialize result to first axis
+        result = _make_elementary_quat(seq[0], angles[:, 0])
+        if len(seq) == 1:
```
I think doing an empty loop is OK as well. I wouldn't add this condition here.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -281,3 +307,113 @@ def as_rotvec(self):
+                # (N, 1)
+                angles = angles[:, None]
+            else:
+                raise ValueError("Expected float or 1D array for parameter "
```
I think having 2D angles with shape (N, 1) is fine as well.

adbugger replied:
You're right. I included it in the doc but forgot to add it here. My mistake.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -281,3 +307,113 @@ def as_rotvec(self):
+                                 "`angles` corresponding to `seq` "
+                                 "got shape {}".format(angles.shape))
+        else:  # 2 or 3 axes
+            if num_axes != angles.shape[-1]:
```
I suggest to do the full check here i.e. if `angles.ndim not in [1, 2] or angles.shape[-1] != num_axes`.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -281,3 +307,113 @@ def as_rotvec(self):
+                             "num_axes), got {}.".format(angles.shape))
+
+        quat = cls._elementary_quat_compose(seq, angles, intrinsic)
+        return cls(quat[0] if is_single else quat)
```
normalized=True?

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -281,3 +307,113 @@ def as_rotvec(self):
+            return result
+
+        for idx, axis in enumerate(seq[1:], start=1):
+            if intrinsic:
```
For intrinsic rotations, shouldn't we do post-multiplication?

nmayorov reviewed (commented):
I like the improvements. Added some more notes.

adbugger committed:
MAINT: Minor changes to calculations and condition checks

adbugger committed:
TST: Elementary test to check extrinsic rotation implementation

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -281,3 +306,113 @@ def as_rotvec(self):
+
+        Parameters
+        ----------
+        seq : str
```
I think it's better to name it 'string', because there are this unicode-str issues in Python.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -281,3 +306,113 @@ def as_rotvec(self):
+        Parameters
+        ----------
+        seq : str
+            `seq` must be a string of upto 3 characters belonging to the set
```
I think you can start simply by `Up to 3 characters ...`.

I would probably prioritize intrinsic rotations --- mention 'X', 'Y', 'Z' first. Also add a reference to wikipedia (probably) about intrinsic and extrinsic rotations.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -281,3 +306,113 @@ def as_rotvec(self):
+            For a single character `seq`, `angles` can be:
+
+                - a single value
+                - array_like, 1D with shape (N, ), where each `angle[i]`
```
I think just `- array_like with shape (N,), ...`. The same goes for other places.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -281,3 +306,113 @@ def as_rotvec(self):
+            and intrinsic rotations cannot be mixed in one function call.
+        angles : float or array_like, shape (N, ) or (N, [1 or 2 or 3])
+            Euler angles specified in radians (`degrees` is False) or degrees
+            if parameter `degrees` is True.
```
Maybe write uniformly ``Euler angles specified in radians (`degrees` is False, default) or degrees (`degrees` is True).``

Or ``Euler angles specified in radians or degrees according to `degrees` arguments.``

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -281,3 +306,113 @@ def as_rotvec(self):
+            represents an extrinsic rotation around the corresponding axis and
+            an uppercase character represents an intrinsic rotation. Extrinsic
+            and intrinsic rotations cannot be mixed in one function call.
+        angles : float or array_like, shape (N, ) or (N, [1 or 2 or 3])
```
I think as you explain possible cases below, you can simply write `array_like`. But I'm fine with your description as well. As you prefer.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -281,3 +306,113 @@ def as_rotvec(self):
+                  corresponding to a sequence of Euler angles describing a
+                  single rotation
+
+        degrees : boolean
```
optional

Default is False in the description for consistency.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -281,3 +306,113 @@ def as_rotvec(self):
+                                 "got shape {}.".format(angles.shape))
+        else:  # 2 or 3 axes
+            if angles.ndim not in [1, 2] or angles.shape[-1] != num_axes:
+                raise ValueError("Expected each angle sequence to have width "
```
The error message should be more general and the `elif` branch below is not necessary anymore.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -281,3 +306,113 @@ def as_rotvec(self):
+
+        # By now angles should have shape (num_rot, num_axes)
+        # sanity check
+        if angles.ndim != 2 or angles.shape[-1] != num_axes:
```
If we are sane programmers, we probably don't need this sanity check. Not sure though.

nmayorov commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -314,3 +314,36 @@ def test_rotvec_calc_pipeline():
+
+
+def test_single_intrinsic_extrinsic_rotation():
+    ext = Rotation.from_euler('z', 90, degrees=True).as_dcm()
```
Just to avoid overwriting `int`, use `extrinsic` and `intrinsic`.

nmayorov commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -314,3 +314,36 @@ def test_rotvec_calc_pipeline():
+    # Intrinsic rotation is same as extrinsic with order reversed
+    np.random.seed(0)
+    a = np.random.randint(low=0, high=180, size=(6, 3))
+    b = a[:, np.argsort([2, 1, 0])]
```
Try `a[:, ::-1]`.

nmayorov commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -314,3 +314,36 @@ def test_rotvec_calc_pipeline():
+    assert_allclose(x, y)
+
+
+def test_from_euler_elementary_extrinsic_rotation():
```
I suggest to test for some angles (like 30, 60, 45 degrees) common rotations 3-1-3 and 3-1-2 or 3-2-1. You can find the DCM for them in many places.

adbugger replied:
Done

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -2,7 +2,31 @@
+    # Einsum calculates row wise dot product
+    product[:, 3] = p[:, 3] * q[:, 3] - np.sum(p[:, :3] * q[:, :3], axis=1)
+    # Vector part of result
+    product[:, :3] = (p[:, 3][:, None] * q[:, :3] +
```
You can write it as `p[:, None, 3]`.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -2,7 +2,31 @@
+    num_quat = p.shape[0]
+    product = np.empty((num_quat, 4))
+    # Scalar part of result
+    # Einsum calculates row wise dot product
```
No einsum left.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -2,7 +2,31 @@
+def _compose_quat(p, q):
+    # p and q should have same shape (N, 4)
+    num_quat = p.shape[0]
+    product = np.empty((num_quat, 4))
```
product = np.empty_like(p)

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -2,7 +2,31 @@
+    num_rotations = angles.shape[0]
+    quat = np.zeros((num_rotations, 4))
+
+    axis_to_ind = {'x': 0, 'y': 1, 'z': 2}
```
I think it's OK to move out this constant dictionary to the module level.

nmayorov reviewed (commented):
Another portion of minor suggestions.

adbugger committed:
MAINT: Changes to `from_euler`

adbugger committed:
TST: Add thorough testing for `from_euler`

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -2,7 +2,46 @@
+import re
+
+
+_axis_to_ind = {'x': 0, 'y': 1, 'z': 2}
```
Use upper case for constants.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -2,7 +2,46 @@
+_axis_to_ind = {'x': 0, 'y': 1, 'z': 2}
+
+
+def _make_elementary_quat(axis, angles):
```
I'm neutral about using underscores here or not using (in all similar places). After all we still can control what to import from `rotation`.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -281,3 +321,105 @@ def as_rotvec(self):
+            intrinsic rotations [1]_, or {'x', 'y', 'z'} for extrinsic
+            rotations [2]_. Extrinsic and intrinsic rotations cannot be mixed
+            in one function call.
+        angles : float or array_like, shape (N, ) or (N, [1 or 2 or 3])
```
I suggest to write `(N,)` without a space (in all similar places).

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -281,3 +321,105 @@ def as_rotvec(self):
+
+        References
+        ----------
+        .. [1] `Intrinsic rotations
```
Can we combine the two references together? It will be more logical and concise.

nmayorov reviewed (commented):
I think these are final and minor notes from me.

Fix how you like it and merge.

adbugger committed:
STY/DOC: Rename constane, combine reference in docstring
